agent/consul/acl_endpoint.go

```diff
@@ -28,6 +28,8 @@ var (
 	validPolicyName              = regexp.MustCompile(`^[A-Za-z0-9\-_]{1,128}$`)
 	validServiceIdentityName     = regexp.MustCompile(`^[a-z0-9]([a-z0-9\-_]*[a-z0-9])?$`)
 	serviceIdentityNameMaxLength = 256
+	validRoleName                = regexp.MustCompile(`^[a-z0-9]([a-z0-9\-_]*[a-z0-9])?$`)
```

Comment:
For existing reasons related to the DNS integration, service names probably should have been validated as valid DNS `labels` as per [rfc1035](https://www.ietf.org/rfc/rfc1035.txt)/[rfc1123](https://www.ietf.org/rfc/rfc1123.txt):

```
<label>       ::= <letter> [ [ <ldh-str> ] <let-dig> ]
<ldh-str>     ::= <let-dig-hyp> | <let-dig-hyp> <ldh-str>
<let-dig-hyp> ::= <let-dig> | "-"
<let-dig>     ::= <letter> | <digit>
<letter>      ::= any one of the 52 alphabetic characters A through Z in upper case
                  and a through z in lower case
<digit>       ::= any one of the ten digits 0 through 9
```

In regex that looks more like `^[a-zA-Z]([a-zA-Z0-9-]*[a-zA-Z0-9])?$` but given dns doesn't care about casing it's practically `^[a-z]([a-z0-9-]*[a-z0-9])?$`.

So technically even my own attempt at this validation is slightly off here, which I should probably correct:
```
^[a-z0-9]([a-z0-9\-_]*[a-z0-9])?$ (this PR)
^[a-z]([a-z0-9-]*[a-z0-9])?$      (DNS label)
```

Introducing this new element (Service Identity) seemed like a reasonable place to correct the errors of old and require a slightly more restrictive character set for services used with this convenience feature.

Furthermore in the context of the secure intro work, the default plumbing path has:

* Kubernetes Service Account Name populates the `ACLToken.Roles[].BoundName` field during login.
* When a role with the bound name does not exist, an effective one is used that maps the `BoundName` to an `ACLServiceIdentity.ServiceName`
* The `ServiceName` field uses an effective policy including `service "<value>" { policy = "write" }`

Given that Kubernetes Service Account Names are similarly\* restrictive, and everything along that chain above (roles & service identity service names) is keyed off of this name it further made sense to restrict these concepts.

Now if we don't actually want to try and correct the errors of old in regards to Service Names and DNS, then the only justification is the Kubernetes one then maybe we just go with something far less restrictive?


(\*) Note: Kubernetes Service Account Names are _nearly_ DNS `labels`. They're actually DNS `subdomains` so they can be a series of `labels` delimited by `.` which is still going to run into Consul DNS issues:
```
$ kubectl create serviceaccount 'FOO_bar'
The ServiceAccount "FOO_bar" is invalid: metadata.name: Invalid value: "FOO_bar": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

Reply:
So if I understand the regex is really to validate a service identity name. The question is why shouldn't I be able to manually create a role called "Admin" or "Security" and instead be forced into lowercase.

Your restrictions make total sense for service identities, just not as much for the lower level role names. I think we should allow names just like we do for policies and have an extra level of verification for configuring service identities.

Reply:
I'll switch it to something similar to the Policy pattern, but with a max length of 256 instead of 128. I'd gone with 256 as it is wide enough to allow someone to template 2 k8s identifiers together (each max length of 63) with enough headroom for some delimiters between them, and some prefix/suffix (then rounded up to a power of two).